### ISSUE TITLE
[Feature request] Ability to set custom display refresh rate 

### ISSUE THREAD

**Issue by eugene-s-nesdev** [punesemu/puNES#150]
Mesen have nice option to force user-specified refresh rate in fullscreen mode.
User can select it from drop-down list of refresh rates that your GPU/Dispaly support: 

![mesen](https://user-images.githubusercontent.com/13199529/120906908-fcc99d00-c665-11eb-8c8f-900c1d450b8c.png)

Nestopia had "auto" refresh rate function in fullscreen depending of region.
For NTSC emulator try to set 60 or 120 Hz, prefer higher value, if hardware supported it.
For PAL/Dendy emulator try to set 50 or 100 Hz, prefer higher value, if hardware supported it.
It was really nice when i had CRT, which supported 60-75-85-100-120Hz 

![nestopia](https://user-images.githubusercontent.com/13199529/120906909-03f0ab00-c666-11eb-889c-491c34114e47.png)

I guess nice idea is combine this methods.
To make an separate "fullscreen preferred refresh rate" option.
Two drop-down menus where user can set prefferred hardware refresh rate: one for NTSC, second for PAL/Dendy

I don't know, though, is really need to allow reading permissible refresh-rates list from system videodriver, or make it fixed?
If some value don't supported by hardware, to let user select another permissible value by yourself.

Fortunately, lastest modern LCD displays have wide range of refresh rates native support, usually 50-144Hz
as well, as AMD FreeSync or Nvidia G-Sync technologies for automatic refresh adjustment.

**Comment by punesemu** [comment]
Hi Eugene, puNES in fullscreen mode use resolution and frequency of the monitor used by the OS. Do you have any flicker problems?

**Comment by eugene-s-nesdev** [comment]
Hi Fabio. No, i don't have problems, all works like a charm.
True reason of this topic/request - ability to smooth (no-tearing) scroll, which is avaliable ONLY
when physical refresh rate corresponding with emulator output.

When emulating NTSC 60Hz games - preferred physical refresh rate = 60Hz or multiple 120Hz
When emulating PAL 50Hz games - preferred physical refresh rate = 50Hz or multiple 100Hz
This settings + vsync gives true smooth scrolling.

Most of PC-hardware runs at default 60Hz refresh rate, so it's good for NTSC but cause jerky scroll on PAL.
Not so much emulators have ability to set preffered refresh rate separately, depending of region.

----

Also, look here please:
https://github.com/libretro/common-shaders/issues/165

**Comment by punesemu** [comment]
> Hi Fabio. No, i don't have problems, all works like a charm.
True reason of this topic/request - ability to smooth (no-tearing) scroll, which is avaliable ONLY
when physical refresh rate corresponding with emulator output.
>
>When emulating NTSC 60Hz games - preferred physical refresh rate = 60Hz or multiple 120Hz
When emulating PAL 50Hz games - preferred physical refresh rate = 50Hz or multiple 100Hz
This settings + vsync gives true smooth scrolling.
>
>Most of PC-hardware runs at default 60Hz refresh rate, so it's good for NTSC but cause jerky scroll on PAL.
Not so much emulators have ability to switch between refresh rates separately depending of region.
Of course, i mean fullscreen mode.

Ok, I've understand, I will try to figure out how to fix the problem. 

>Also, look here please:
libretro/common-shaders#165
>
>Your last commit 31a7bd2
fixes the crash problem. Do you saw that topic before?

No, I hadn't seen the discussion, it was just a coincidence 😄. I had downloaded the latest shaders and found that some did not work giving me black screen or crashing or loading the variables wrong and I wanted to eliminate all the problems.

**Comment by punesemu** [comment]
Hi Eugene, I've implemented the feature but for now only under Linux. Do
you have the possibilities to test? I don't have a monitor that support 50
Hz or 100 Hz so i can't test. I've also implemented the option to choose
the resolution for the fullscreen but this is also to be tested. If you
give me confirmation I will send you the patch

Fabio

Il giorno sab 19 giu 2021 alle ore 16:30 eugene-s-nesdev <
***@***.***> ha scritto:
